#!/home/zhengjh/miniconda3/bin/Rscript
# parameter

library(optparse)
library(getopt)

option_list <- list(
  
  make_option(c("-o", "--output"), type = "character", default = FALSE,
              action = "store", help = "This is the output directory."
  ),
  make_option(c("-t", "--Type"), type = "character", default = "10X",
              action = "store", help = "This is type of sequencing library, default is 10X"
  ),
  make_option(c("--path_normal"), type = "character", default = FALSE,
              action = "store", help = "This is the path of normal exp matrix."
  ),
  make_option(c("--path_treat"), type = "character", default = FALSE,
              action = "store", help = "This is the path of treat exp matrix."
  ),
  make_option(c("--folder_name"), type = "character", default = "unfilter",
              action = "store", help = "This is the folder name of results, default is the unfiler, if you want to filter, you need to change it"
  ),
  make_option(c("--project_name"), type = "character", default = FALSE,
              action = "store", help = "This is project name built in seurat object and the pre_fix of files generated by this function."
  ),
  make_option(c("--mt_pattern"), type = "character", default = "^mt-",
              action = "store", help = "This is mitochondria pattern used to calculate its percentage, default is the ^mt-."
  ),
  make_option(c("--rp_pattern"), type = "character", default = "^Rpl|^Rps",
              action = "store", help = "This is ribosome pattern used to calculate its percentage, default is ^Rpl|^Rps"
  ),
  
  make_option(c("--min_cell"), type = "integer", default = 3,
              action = "store", help = "This is min cells number to filter, default is 3."
  ),
  make_option(c("--min_gene"), type = "integer", default = 0,
              action = "store", help = "This is min gene number to filter, default is 0."
  ),
  
  make_option(c("--min_ncountRNA"), type = "integer", default = 0,
              action = "store", help = "This is min ncountRNA to filter, default it 0."
  ),
  make_option(c("--min_mt_percent"), type = "integer", default = 40,
              action = "store", help = "This is min mitochondria percentage  to filter, default is 40."
  ),
  make_option(c("--min_rp_percent"), type = "integer", default = 20,
              action = "store", help = "This is min ribosome percentage to filter, default is 20."
  ),
  make_option(c("--npc_plot"), type = "integer", default = 50,
              action = "store", help = "This is number of principles  to plot, default is 50."
  ),
  make_option(c("--npc_used"), type = "integer", default = 20,
              action = "store", help = "This is number of principles to used, default is 20."
  ),
  make_option(c("--resolution_number"), type = "double", default = 0.5,
              action = "store", help = "This is resolution number to used, default is 0.5"
  ),
  make_option(c("--k_parameter"), type = "integer", default = 20,
              action = "store", help = "This is k parameter  to used which will influence the clusering number, default is 20."
  )
)

# -help 
opt = parse_args(OptionParser(option_list = option_list, usage = "This Script is general processing of single cell exp matrix!"))
print(opt)

##--step00. load functions and ref objects for cell type transfer
# read in the functions
timestart<-Sys.time() 
source("/home/zhengjh/bin/Code/Seurat_function.R")
source("/home/zhengjh/bin/Code/diffprop_functions.R")

# set the output path
setwd(opt$output)
# load ref object
ref_object_Seurat <- readRDS("/home/zhengjh/projects/Ob/Data/Cell_type_identification_data/ExoExp_RefSeuratObject/GSE113576_ref_seurat.rds")
##--step01. load exp matrix
if(opt$Type=="10X"){
  normal_exp <- Read10X(data.dir = opt$path_normal)
  print("The rownum and colnum of normal exp is ")
  print(dim(normal_exp))
  treat_exp <- Read10X(data.dir = opt$path_treat) 
  print("The rownum and colnum of treat exp is ")
  print(dim(treat_exp))
}else{
  normal_exp <- read.csv(opt$path_normal, header = T, row.names = 1) 
  print("The rownum and colnum of normal exp is ")
  print(dim(normal_exp))
  treat_exp <- read.csv(opt$path_treat, header = T, row.names = 1) 
  print("The rownum and colnum of treat exp is ")
  print(dim(treat_exp))
}

##--step1: Create_seurat_filter_cell and QC
if(opt$folder_name=="unfilter"){
  exp_count <- cbind(normal_exp, treat_exp)
  n_normal <- dim(normal_exp)[2]
  n_treat <- dim(treat_exp)[2]
  Or_ident <- c(rep("Control", n_normal), rep("Obesity",n_treat))
  Aggregated_seurat <- Create_seurat_filter_cell(exp_count, opt$folder_name, opt$project_name, opt$min_cell, Or_ident, opt$mt_pattern, opt$rp_pattern)
  print("The row_num and col_num of the original matrix:")
  print(dim(Aggregated_seurat))
  print("The cellnumber of samples:")
  print(table(Aggregated_seurat$orig.ident))
  name_pre <- paste0("filter_gene_expressed_", opt$min_cell,"cells") 
  QC(Aggregated_seurat,name_pre)
}else{
  ## original
  exp_count <- cbind(normal_exp, treat_exp)
  n_normal <- dim(normal_exp)[2]
  n_treat <- dim(treat_exp)[2]
  Or_ident <- c(rep("Control", n_normal), rep("Obesity",n_treat))
  Aggregated_seurat <- Create_seurat_filter_cell(exp_count,opt$folder_name, opt$project_name, opt$min_cell, Or_ident, opt$mt_pattern, opt$rp_pattern)
  print("The row_num and col_num of the original matrix:")
  print(dim(Aggregated_seurat))
  print("The cellnumber of samples:")
  print(table(Aggregated_seurat$orig.ident))
  name_pre <- paste0("filter_gene_expressed_", opt$min_cell,"cells") 
  QC(Aggregated_seurat,name_pre)
  ## filter
  Aggregated_seurat <- Subset_data_QC(Aggregated_seurat,opt$min_gene,opt$min_mt_percent,opt$min_rp_percent,opt$min_ncountRNA)
  print("The row_num and col_num of the filtering matrix:")
  print(dim(Aggregated_seurat))
  print("The cellnumber of filtering samples:")
  print(table(Aggregated_seurat$orig.ident))
}

##--step2: Normalize Find VariableGene and Scale Data and then RunPCA
Aggregated_seurat <- Select_pc(Aggregated_seurat, opt$project_name, opt$npc_plot)
##--step3: RunUMAP and RunTSNE and Then find cluster
Aggregated_seurat <- Plot_cluster(Aggregated_seurat, opt$project_name, opt$npc_used, opt$resolution_number, opt$k_parameter)
##--step4: DEGs between cluster 
Aggregated_seurat_markers <- FindmarkerForCluster(Aggregated_seurat, opt$project_name)
write.csv(Aggregated_seurat_markers, paste0(opt$project_name,"_DEGsbetweenClusters.csv"))

##--step5: Topmarker of each cluster
top_num <- 10
TopMarkersInClusters <- TopMarkersInCluster(Aggregated_seurat_markers, opt$project_name, top_num)
# Top10 markers heatmap
P_heatmap_top_marker <- DoHeatmap(Aggregated_seurat, features = TopMarkersInClusters$gene)
pdf(paste0(opt$project_name, "Heatmap_plot_top10_markers.pdf"), 30, 15)
print(P_heatmap_top_marker)
dev.off()

##--step6: celltype identification-----------------------------
ref.seurat <- ref_object_Seurat
Aggregated_seurat <- Classification_Seurat(Aggregated_seurat, ref.seurat)

metadata_names <- colnames(Aggregated_seurat@meta.data)
remove_index <- grep(pattern="prediction", metadata_names)
remove_name <- metadata_names[remove_index]
Aggregated_seurat@meta.data <- Aggregated_seurat@meta.data[, -which(metadata_names %in% remove_name)]
write.csv(Aggregated_seurat@meta.data, paste0(opt$project_name,"_metadata.csv"))

##--step7: featureplot of marker genes
markers <- c("Agt","Flt1","Ccdc153","Mia","Rax",
             "Col1a2","Pdgfra","Mrc1","Mobp",
             "Cx3cr1","Vtn","Snap25","Syt1","Fyn")
p <- FeaturePlot(Aggregated_seurat, features = markers,
                 reduction = "umap", ncol = 4, col = c("lightgrey", "red"))
pdf(paste0(opt$project_name,"_featureplot_plot_marker.pdf"), 22, 16)
print(p)
dev.off()

##--step8: save the clustering plots aftering assign celltypes
p1 <- DimPlot(Aggregated_seurat, reduction = "umap",label = TRUE,
              pt.size = 0.8, label.size = 4.5, repel = TRUE, group.by = "ident") + NoLegend()
p1 <- p1 + labs(title="Clustering")
p1 <- p1 + theme(plot.title = element_text(hjust = 0.5)) 
p2 <- DimPlot(Aggregated_seurat, reduction = "umap", label = TRUE,
              pt.size = 0.8, label.size = 4.5, repel = TRUE, group.by="celltype_Seurat")
p2  <- p2  + labs(title="Identification_celltype_by_Seurat_lable_transfer") + theme(plot.title = element_text(hjust = 0.5)) 

p3 <- DimPlot(Aggregated_seurat, reduction = "umap",label = TRUE,
              pt.size = 0.8, label.size = 4.5, repel = TRUE, group.by = "ident") + NoLegend()
p3 <- p3 + labs(title="Clustering")
p3 <- p3 + theme(plot.title = element_text(hjust = 0.5)) 
p4 <- DimPlot(Aggregated_seurat, reduction = "umap", label = TRUE,
              pt.size = 0.8, label.size = 4.5, repel = TRUE, group.by="celltype_Seurat")
p4  <- p4  + labs(title="Identification_celltype_by_Seurat_lable_transfer") + theme(plot.title = element_text(hjust = 0.5)) 

plots_tsne <- plot_grid(p1, p2, ncol=2,rel_widths = c(1.5,2))
plots_umap <- plot_grid(p3, p4, ncol=2,rel_widths = c(1.5,2))
pdf("Clustering_celltype_by_Seurat_transfer_fun.pdf",18,6)
print(plots_tsne)
print(plots_umap)
dev.off()

rds_name <- paste0(opt$project_name, "_Aggregated_seurat.rds")
saveRDS(Aggregated_seurat, file = rds_name)

timeend<-Sys.time()
runningtime<-timeend-timestart
print(runningtime)
